File: packages/checkin-core/src/checkin.test.ts
import { test, describe } from "node:test";
import assert from "node:assert/strict";
import { randomBytes, randomUUID } from "node:crypto";

import { issueToken, verifyToken, type EventKey } from "./token.ts";
import { decide, type Context, type LocalInvitation, type Rsvp } from "./checkin.ts";

// ---------------------------------------------------------------------------
// fixtures — Ahmed & Aisha, Lagos leg
// ---------------------------------------------------------------------------

const WEDDING = randomUUID();
const OTHER_WEDDING = randomUUID();
const LAGOS_LEG = randomUUID();

const weddingKey: EventKey = {
  eventId: WEDDING,
  eventName: "Ahmed & Aisha",
  tokenVersion: 1,
  key: randomBytes(32),
};
const otherKey: EventKey = {
  eventId: OTHER_WEDDING,
  eventName: "Yusuf & Maryam",
  tokenVersion: 1,
  key: randomBytes(32),
};

function household(over: Partial<LocalInvitation> = {}): LocalInvitation {
  return {
    passId: randomUUID(),
    invitationId: randomUUID(),
    eventId: WEDDING,
    legId: LAGOS_LEG,
    displayName: "Mr & Mrs Adeyemi",
    category: "Groom's Family",
    tableName: "Table 12",
    allowance: 4,
    admitted: 0,
    rsvp: "attending",
    revoked: false,
    ...over,
  };
}

function ctx(list: LocalInvitation[], over: Partial<Context> = {}): Context {
  const byId = new Map(list.map((i) => [i.passId, i]));
  return {
    currentEventId: WEDDING,
    currentLegId: LAGOS_LEG,
    policy: { allowOverflow: true, requireRsvp: false },
    keys: [weddingKey, otherKey],
    find: (id) => byId.get(id),
    canOverrideRsvp: false,
    ...over,
  };
}

const tokenFor = (inv: LocalInvitation, key = weddingKey, version = 1) =>
  issueToken({ passId: inv.passId, eventId: inv.eventId, tokenVersion: version }, key.key);

// ---------------------------------------------------------------------------

describe("token", () => {
  test("round trips and stays short enough for a low-density QR", () => {
    const inv = household();
    const t = tokenFor(inv);
    assert.ok(t.length < 80, `token was ${t.length} chars`);

    const v = verifyToken(t, [weddingKey]);
    assert.ok(v.ok);
    assert.equal(v.payload.passId, inv.passId);
    assert.equal(v.payload.eventId, WEDDING);
  });

  test("rejects a tampered payload", () => {
    const inv = household();
    const t = tokenFor(inv);
    const parts = t.split(".");
    // Replace the last char with one that differs from it — a quarter of
    // random uuids already end in "A", and "forging" a token into itself
    // is no forgery at all.
    const swap = parts[0].endsWith("A") ? "B" : "A";
    const forged = [parts[0].slice(0, -1) + swap, parts[1], parts[2], parts[3]].join(".");
    assert.equal(verifyToken(forged, [weddingKey]).ok, false);
  });

  test("rejects a signature from an unknown key", () => {
    const stranger = randomBytes(32);
    const t = issueToken({ passId: randomUUID(), eventId: WEDDING, tokenVersion: 1 }, stranger);
    assert.equal(verifyToken(t, [weddingKey]).ok, false);
  });

  test("identifies which held event a foreign pass belongs to", () => {
    const t = issueToken(
      { passId: randomUUID(), eventId: OTHER_WEDDING, tokenVersion: 1 },
      otherKey.key,
    );
    const v = verifyToken(t, [weddingKey, otherKey]);
    assert.ok(v.ok);
    assert.equal(v.matched.eventName, "Yusuf & Maryam");
  });

  test("a version bump kills every old pass at once", () => {
    const inv = household();
    const old = tokenFor(inv, weddingKey, 1);
    const reissued: EventKey = { ...weddingKey, tokenVersion: 2 };
    const v = verifyToken(old, [reissued]);
    assert.equal(v.ok, false);
    assert.equal((v as { reason: string }).reason, "stale_version");
  });

  test("garbage in, malformed out", () => {
    for (const junk of ["", "hello", "a.b.c", "a.b.c.d.e", "....", "https://example.com"]) {
      assert.equal(verifyToken(junk, [weddingKey]).ok, false);
    }
  });
});

// ---------------------------------------------------------------------------

describe("admitting", () => {
  test("a single guest goes straight through with no prompt", () => {
    const inv = household({ allowance: 1, displayName: "Chidinma Okafor", tableName: "Table 7" });
    const d = decide(ctx([inv]), { kind: "scan", raw: tokenFor(inv) });

    assert.equal(d.outcome, "admitted");
    assert.equal(d.tone, "admit");
    assert.equal(d.admittedCount, 1);
    assert.equal(d.remaining, 0);
    assert.equal(d.autoReturnMs, 1500, "must return to camera on its own");
    assert.ok(d.log);
  });

  test("a party of four is asked how many arrived", () => {
    const inv = household();
    const d = decide(ctx([inv]), { kind: "scan", raw: tokenFor(inv) });

    assert.equal(d.outcome, "needs_count");
    assert.equal(d.tone, "ask");
    assert.deepEqual(d.choices, [1, 2, 3, 4]);
    assert.equal(d.remaining, 4);
    assert.equal(d.log, false, "the prompt itself is not a log row");
    assert.equal(d.autoReturnMs, null);
  });

  test("three of four is a partial admit and the pass stays live", () => {
    const inv = household();
    const d = decide(ctx([inv]), { kind: "scan", raw: tokenFor(inv), requestedCount: 3 });

    assert.equal(d.outcome, "partial");
    assert.equal(d.tone, "admit");
    assert.equal(d.admittedCount, 3);
    assert.equal(d.remaining, 1);
    assert.equal(d.autoReturnMs, 2500, "longer dwell — there is a number to read");
  });

  test("the fourth arriving later is admitted, not treated as a duplicate", () => {
    const inv = household({ admitted: 3 });
    const first = decide(ctx([inv]), { kind: "scan", raw: tokenFor(inv) });

    assert.equal(first.outcome, "needs_count");
    assert.deepEqual(first.choices, [1], "only one left to offer");

    const d = decide(ctx([inv]), { kind: "scan", raw: tokenFor(inv), requestedCount: 1 });
    assert.equal(d.outcome, "admitted");
    assert.equal(d.remaining, 0);
  });

  test("taking all four at once completes the party", () => {
    const inv = household();
    const d = decide(ctx([inv]), { kind: "scan", raw: tokenFor(inv), requestedCount: 4 });
    assert.equal(d.outcome, "admitted");
    assert.equal(d.remaining, 0);
  });
});

// ---------------------------------------------------------------------------

describe("allowance exhausted", () => {
  test("with overflow off, a fully admitted party is held", () => {
    const inv = household({ admitted: 4 });
    const d = decide(ctx([inv], { policy: { allowOverflow: false, requireRsvp: false } }), {
      kind: "scan",
      raw: tokenFor(inv),
    });

    assert.equal(d.outcome, "allowance_exhausted");
    assert.equal(d.tone, "hold");
    assert.equal(d.admittedCount, 0);
    assert.equal(d.autoReturnMs, null, "someone is talking at the gate — do not reset");
    assert.match(d.detail!, /4 of 4/);
  });

  test("with overflow on, the usher is offered the choice instead", () => {
    const inv = household({ admitted: 4 });
    const d = decide(ctx([inv]), { kind: "scan", raw: tokenFor(inv) });

    assert.equal(d.outcome, "needs_count");
    assert.equal(d.remaining, 0);
    assert.match(d.detail!, /over the invitation/);
  });
});

// ---------------------------------------------------------------------------

describe("overflow", () => {
  test("five through a party of four is admitted and flagged", () => {
    const inv = household({ displayName: "The Nwosu Family" });
    const d = decide(ctx([inv]), { kind: "scan", raw: tokenFor(inv), requestedCount: 5 });

    assert.equal(d.outcome, "overflow_admitted");
    assert.equal(d.tone, "hold", "amber, not green — it is a decision, not a routine admit");
    assert.equal(d.admittedCount, 5, "everyone standing there gets in");
    assert.match(d.detail!, /organiser has been notified/);
  });

  test("blocked when the organiser turned overflow off", () => {
    const inv = household();
    const d = decide(ctx([inv], { policy: { allowOverflow: false, requireRsvp: false } }), {
      kind: "scan",
      raw: tokenFor(inv),
      requestedCount: 5,
    });

    assert.equal(d.outcome, "overflow_blocked");
    assert.equal(d.tone, "deny");
    assert.equal(d.admittedCount, 0);
    assert.ok(d.actions.includes("Call manager"));
  });
});

// ---------------------------------------------------------------------------

describe("refusing", () => {
  test("a revoked pass names the household", () => {
    const inv = household({ revoked: true, displayName: "Tunde Bakare" });
    const d = decide(ctx([inv]), { kind: "scan", raw: tokenFor(inv) });

    assert.equal(d.outcome, "revoked");
    assert.equal(d.tone, "deny");
    assert.match(d.detail!, /Tunde Bakare/);
  });

  test("a pass from another wedding says which one", () => {
    const foreign = issueToken(
      { passId: randomUUID(), eventId: OTHER_WEDDING, tokenVersion: 1 },
      otherKey.key,
    );
    const d = decide(ctx([]), { kind: "scan", raw: foreign });

    assert.equal(d.outcome, "wrong_event");
    assert.match(d.detail!, /Yusuf & Maryam/, "this is the whole point of multi-key loading");
  });

  test("a genuine pass for the other leg is distinguished from a forgery", () => {
    const abujaOnly = household();
    // Signed for this event, but absent from the Lagos leg's list.
    const d = decide(ctx([]), { kind: "scan", raw: tokenFor(abujaOnly) });

    assert.equal(d.outcome, "wrong_leg");
    assert.notEqual(d.outcome, "invalid");
  });

  test("an unreadable code offers name search rather than dead-ending", () => {
    const d = decide(ctx([]), { kind: "scan", raw: "not-a-token" });

    assert.equal(d.outcome, "invalid");
    assert.ok(d.actions.includes("Search by name"), "usually a real guest with a broken screen");
  });

  test("a reissued pass reads as revoked, not invalid", () => {
    const inv = household();
    const old = tokenFor(inv, weddingKey, 1);
    const d = decide(ctx([inv], { keys: [{ ...weddingKey, tokenVersion: 2 }] }), {
      kind: "scan",
      raw: old,
    });

    assert.equal(d.outcome, "revoked");
    assert.match(d.detail!, /new link/);
  });
});

// ---------------------------------------------------------------------------

describe("the RSVP gate", () => {
  const cases: [Rsvp, boolean, string][] = [
    ["pending", false, "admitted"],
    ["pending", true, "rsvp_blocked"],
    ["attending", true, "needs_count"],
    ["partial", true, "needs_count"],
  ];

  for (const [rsvp, requireRsvp, expected] of cases) {
    test(`rsvp=${rsvp}, required=${requireRsvp} → ${expected}`, () => {
      const inv = household({ rsvp, allowance: rsvp === "pending" ? 1 : 4 });
      const d = decide(ctx([inv], { policy: { allowOverflow: true, requireRsvp } }), {
        kind: "scan",
        raw: tokenFor(inv),
      });
      assert.equal(d.outcome, expected);
    });
  }

  test("declining is refused whatever the policy says", () => {
    const inv = household({ rsvp: "declined" });
    const d = decide(ctx([inv], { policy: { allowOverflow: true, requireRsvp: false } }), {
      kind: "scan",
      raw: tokenFor(inv),
    });
    assert.equal(d.outcome, "rsvp_declined");
  });

  test("override is offered only to ushers who hold the permission", () => {
    const inv = household({ rsvp: "declined" });

    const without = decide(ctx([inv]), { kind: "scan", raw: tokenFor(inv) });
    assert.ok(!without.actions.includes("Admit anyway"));

    const with_ = decide(ctx([inv], { canOverrideRsvp: true }), {
      kind: "scan",
      raw: tokenFor(inv),
    });
    assert.ok(with_.actions.includes("Admit anyway"));
  });
});

// ---------------------------------------------------------------------------

describe("manual check-in", () => {
  test("skips every token check and is logged as manual", () => {
    const inv = household({ allowance: 1 });
    const d = decide(ctx([inv]), { kind: "manual", passId: inv.passId, requestedCount: 1 });

    assert.equal(d.outcome, "manual");
    assert.equal(d.tone, "admit");
    assert.equal(d.admittedCount, 1);
  });

  test("a name that is not on the list offers a walk-in", () => {
    const d = decide(ctx([]), { kind: "manual", passId: randomUUID() });

    assert.equal(d.outcome, "not_found");
    assert.ok(d.actions.includes("Add walk-in"));
  });
});

// ---------------------------------------------------------------------------

describe("invariants that must hold everywhere", () => {
  const everyScenario = () => {
    const good = household();
    const single = household({ allowance: 1 });
    const done = household({ admitted: 4 });
    const gone = household({ revoked: true });
    const no = household({ rsvp: "declined" });

    return [
      decide(ctx([single]), { kind: "scan", raw: tokenFor(single) }),
      decide(ctx([good]), { kind: "scan", raw: tokenFor(good) }),
      decide(ctx([good]), { kind: "scan", raw: tokenFor(good), requestedCount: 2 }),
      decide(ctx([good]), { kind: "scan", raw: tokenFor(good), requestedCount: 9 }),
      decide(ctx([done], { policy: { allowOverflow: false, requireRsvp: false } }), {
        kind: "scan",
        raw: tokenFor(done),
      }),
      decide(ctx([gone]), { kind: "scan", raw: tokenFor(gone) }),
      decide(ctx([no]), { kind: "scan", raw: tokenFor(no) }),
      decide(ctx([]), { kind: "scan", raw: "rubbish" }),
      decide(ctx([]), { kind: "manual", passId: randomUUID() }),
    ];
  };

  test("only green outcomes reset the camera by themselves", () => {
    for (const d of everyScenario()) {
      if (d.autoReturnMs !== null) {
        assert.equal(d.tone === "admit" || d.outcome === "overflow_admitted", true,
          `${d.outcome} auto-returned but is not an admission`);
      }
    }
  });

  test("nothing is admitted on a refusal", () => {
    const refusals = new Set([
      "invalid", "wrong_event", "wrong_leg", "revoked",
      "rsvp_blocked", "rsvp_declined", "allowance_exhausted",
      "overflow_blocked", "not_found",
    ]);
    for (const d of everyScenario()) {
      if (refusals.has(d.outcome)) assert.equal(d.admittedCount, 0);
    }
  });

  test("every refusal is recorded — the organiser wants that report", () => {
    for (const d of everyScenario()) {
      if (d.tone === "deny" || d.tone === "hold") assert.ok(d.log, `${d.outcome} was not logged`);
    }
  });

  test("only the count prompt goes unlogged", () => {
    for (const d of everyScenario()) {
      if (!d.log) assert.equal(d.outcome, "needs_count");
    }
  });

  test("no refusal leaks another household's identity", () => {
    const foreign = issueToken(
      { passId: randomUUID(), eventId: OTHER_WEDDING, tokenVersion: 1 },
      otherKey.key,
    );
    const d = decide(ctx([]), { kind: "scan", raw: foreign });
    assert.equal(d.invitation, undefined);
  });

  test("every outcome gives the usher something to read", () => {
    for (const d of everyScenario()) {
      assert.ok(d.headline.length > 0);
    }
  });
});

// ---------------------------------------------------------------------------

describe("two phones, both offline, same pass", () => {
  test("each admits independently — reconciliation is the server's job", () => {
    const shared = household({ allowance: 4 });

    // Neither device has seen the other's scan.
    const mainGate = decide(ctx([{ ...shared }]), {
      kind: "scan",
      raw: tokenFor(shared),
      requestedCount: 3,
    });
    const sideGate = decide(ctx([{ ...shared }]), {
      kind: "scan",
      raw: tokenFor(shared),
      requestedCount: 2,
    });

    assert.equal(mainGate.admittedCount, 3);
    assert.equal(sideGate.admittedCount, 2);

    // Five admitted against an allowance of four. Both rows are kept, the
    // server flags the overlap, and nobody already inside is thrown out.
    assert.equal(mainGate.admittedCount + sideGate.admittedCount, 5);
    assert.ok(mainGate.log && sideGate.log);
  });
});

File: apps/scanner/test/checkin_test.dart
// Dart port of packages/checkin-core/src/checkin.test.ts — all 35 cases.
// The suites and assertions mirror the TypeScript original 1:1; if the two
// files ever disagree, both are wrong.
import 'dart:math';

import 'package:flutter_test/flutter_test.dart';
import 'package:scanner/core/checkin.dart';
import 'package:scanner/core/token.dart';

// ---------------------------------------------------------------------------
// fixtures — Ahmed & Aisha, Lagos leg
// ---------------------------------------------------------------------------

final _rng = Random.secure();

List<int> randomBytes(int n) => List<int>.generate(n, (_) => _rng.nextInt(256));

String randomUuid() {
  final b = randomBytes(16);
  b[6] = (b[6] & 0x0f) | 0x40;
  b[8] = (b[8] & 0x3f) | 0x80;
  final h = b.map((x) => x.toRadixString(16).padLeft(2, '0')).join();
  return '${h.substring(0, 8)}-${h.substring(8, 12)}-${h.substring(12, 16)}-'
      '${h.substring(16, 20)}-${h.substring(20)}';
}

final wedding = randomUuid();
final otherWedding = randomUuid();
final lagosLeg = randomUuid();

final weddingKey = EventKey(
  eventId: wedding,
  eventName: 'Ahmed & Aisha',
  tokenVersion: 1,
  key: randomBytes(32),
);
final otherKey = EventKey(
  eventId: otherWedding,
  eventName: 'Yusuf & Maryam',
  tokenVersion: 1,
  key: randomBytes(32),
);

LocalInvitation household({
  String? passId,
  String? displayName,
  String? tableName,
  int allowance = 4,
  int admitted = 0,
  Rsvp rsvp = Rsvp.attending,
  bool revoked = false,
}) =>
    LocalInvitation(
      passId: passId ?? randomUuid(),
      invitationId: randomUuid(),
      eventId: wedding,
      legId: lagosLeg,
      displayName: displayName ?? 'Mr & Mrs Adeyemi',
      category: "Groom's Family",
      tableName: tableName ?? 'Table 12',
      allowance: allowance,
      admitted: admitted,
      rsvp: rsvp,
      revoked: revoked,
    );

Context ctx(
  List<LocalInvitation> list, {
  Policy? policy,
  List<EventKey>? keys,
  bool canOverrideRsvp = false,
}) {
  final byId = {for (final i in list) i.passId: i};
  return Context(
    currentEventId: wedding,
    currentLegId: lagosLeg,
    policy: policy ?? const Policy(allowOverflow: true, requireRsvp: false),
    keys: keys ?? [weddingKey, otherKey],
    find: (id) => byId[id],
    canOverrideRsvp: canOverrideRsvp,
  );
}

String tokenFor(LocalInvitation inv, {EventKey? key, int version = 1}) =>
    issueToken(
      TokenPayload(
        passId: inv.passId,
        eventId: inv.eventId,
        tokenVersion: version,
      ),
      (key ?? weddingKey).key,
    );

// ---------------------------------------------------------------------------

void main() {
  group('token', () {
    test('round trips and stays short enough for a low-density QR', () {
      final inv = household();
      final t = tokenFor(inv);
      expect(t.length, lessThan(80), reason: 'token was ${t.length} chars');

      final v = verifyToken(t, [weddingKey]);
      expect(v.ok, isTrue);
      final ok = v as VerifyOk;
      expect(ok.payload.passId, inv.passId);
      expect(ok.payload.eventId, wedding);
    });

    test('rejects a tampered payload', () {
      final inv = household();
      final t = tokenFor(inv);
      final parts = t.split('.');
      // Replace the last char with one that differs from it — a quarter of
      // random uuids already end in "A", and "forging" a token into itself
      // is no forgery at all.
      final swap = parts[0].endsWith('A') ? 'B' : 'A';
      final forged = [
        parts[0].substring(0, parts[0].length - 1) + swap,
        parts[1],
        parts[2],
        parts[3],
      ].join('.');
      expect(verifyToken(forged, [weddingKey]).ok, isFalse);
    });

    test('rejects a signature from an unknown key', () {
      final stranger = randomBytes(32);
      final t = issueToken(
        TokenPayload(passId: randomUuid(), eventId: wedding, tokenVersion: 1),
        stranger,
      );
      expect(verifyToken(t, [weddingKey]).ok, isFalse);
    });

    test('identifies which held event a foreign pass belongs to', () {
      final t = issueToken(
        TokenPayload(
            passId: randomUuid(), eventId: otherWedding, tokenVersion: 1),
        otherKey.key,
      );
      final v = verifyToken(t, [weddingKey, otherKey]);
      expect(v.ok, isTrue);
      expect((v as VerifyOk).matched.eventName, 'Yusuf & Maryam');
    });

    test('a version bump kills every old pass at once', () {
      final inv = household();
      final old = tokenFor(inv, version: 1);
      final reissued = EventKey(
        eventId: weddingKey.eventId,
        eventName: weddingKey.eventName,
        tokenVersion: 2,
        key: weddingKey.key,
      );
      final v = verifyToken(old, [reissued]);
      expect(v.ok, isFalse);
      expect((v as VerifyFail).reason, 'stale_version');
    });

    test('garbage in, malformed out', () {
      for (final junk in [
        '',
        'hello',
        'a.b.c',
        'a.b.c.d.e',
        '....',
        'https://example.com',
      ]) {
        expect(verifyToken(junk, [weddingKey]).ok, isFalse);
      }
    });
  });

  // -------------------------------------------------------------------------

  group('admitting', () {
    test('a single guest goes straight through with no prompt', () {
      final inv = household(
          allowance: 1, displayName: 'Chidinma Okafor', tableName: 'Table 7');
      final d = decide(ctx([inv]), ScanRaw(tokenFor(inv)));

      expect(d.outcome, Outcome.admitted);
      expect(d.tone, Tone.admit);
      expect(d.admittedCount, 1);
      expect(d.remaining, 0);
      expect(d.autoReturnMs, 1500,
          reason: 'must return to camera on its own');
      expect(d.log, isTrue);
    });

    test('a party of four is asked how many arrived', () {
      final inv = household();
      final d = decide(ctx([inv]), ScanRaw(tokenFor(inv)));

      expect(d.outcome, Outcome.needsCount);
      expect(d.tone, Tone.ask);
      expect(d.choices, [1, 2, 3, 4]);
      expect(d.remaining, 4);
      expect(d.log, isFalse, reason: 'the prompt itself is not a log row');
      expect(d.autoReturnMs, isNull);
    });

    test('three of four is a partial admit and the pass stays live', () {
      final inv = household();
      final d = decide(ctx([inv]), ScanRaw(tokenFor(inv), requestedCount: 3));

      expect(d.outcome, Outcome.partial);
      expect(d.tone, Tone.admit);
      expect(d.admittedCount, 3);
      expect(d.remaining, 1);
      expect(d.autoReturnMs, 2500,
          reason: 'longer dwell — there is a number to read');
    });

    test('the fourth arriving later is admitted, not treated as a duplicate',
        () {
      final inv = household(admitted: 3);
      final first = decide(ctx([inv]), ScanRaw(tokenFor(inv)));

      expect(first.outcome, Outcome.needsCount);
      expect(first.choices, [1], reason: 'only one left to offer');

      final d = decide(ctx([inv]), ScanRaw(tokenFor(inv), requestedCount: 1));
      expect(d.outcome, Outcome.admitted);
      expect(d.remaining, 0);
    });

    test('taking all four at once completes the party', () {
      final inv = household();
      final d = decide(ctx([inv]), ScanRaw(tokenFor(inv), requestedCount: 4));
      expect(d.outcome, Outcome.admitted);
      expect(d.remaining, 0);
    });
  });

  // -------------------------------------------------------------------------

  group('allowance exhausted', () {
    test('with overflow off, a fully admitted party is held', () {
      final inv = household(admitted: 4);
      final d = decide(
        ctx([inv],
            policy: const Policy(allowOverflow: false, requireRsvp: false)),
        ScanRaw(tokenFor(inv)),
      );

      expect(d.outcome, Outcome.allowanceExhausted);
      expect(d.tone, Tone.hold);
      expect(d.admittedCount, 0);
      expect(d.autoReturnMs, isNull,
          reason: 'someone is talking at the gate — do not reset');
      expect(d.detail, contains('4 of 4'));
    });

    test('with overflow on, the usher is offered the choice instead', () {
      final inv = household(admitted: 4);
      final d = decide(ctx([inv]), ScanRaw(tokenFor(inv)));

      expect(d.outcome, Outcome.needsCount);
      expect(d.remaining, 0);
      expect(d.detail, contains('over the invitation'));
    });
  });

  // -------------------------------------------------------------------------

  group('overflow', () {
    test('five through a party of four is admitted and flagged', () {
      final inv = household(displayName: 'The Nwosu Family');
      final d = decide(ctx([inv]), ScanRaw(tokenFor(inv), requestedCount: 5));

      expect(d.outcome, Outcome.overflowAdmitted);
      expect(d.tone, Tone.hold,
          reason: 'amber, not green — it is a decision, not a routine admit');
      expect(d.admittedCount, 5,
          reason: 'everyone standing there gets in');
      expect(d.detail, contains('organiser has been notified'));
    });

    test('blocked when the organiser turned overflow off', () {
      final inv = household();
      final d = decide(
        ctx([inv],
            policy: const Policy(allowOverflow: false, requireRsvp: false)),
        ScanRaw(tokenFor(inv), requestedCount: 5),
      );

      expect(d.outcome, Outcome.overflowBlocked);
      expect(d.tone, Tone.deny);
      expect(d.admittedCount, 0);
      expect(d.actions, contains('Call manager'));
    });
  });

  // -------------------------------------------------------------------------

  group('refusing', () {
    test('a revoked pass names the household', () {
      final inv = household(revoked: true, displayName: 'Tunde Bakare');
      final d = decide(ctx([inv]), ScanRaw(tokenFor(inv)));

      expect(d.outcome, Outcome.revoked);
      expect(d.tone, Tone.deny);
      expect(d.detail, contains('Tunde Bakare'));
    });

    test('a pass from another wedding says which one', () {
      final foreign = issueToken(
        TokenPayload(
            passId: randomUuid(), eventId: otherWedding, tokenVersion: 1),
        otherKey.key,
      );
      final d = decide(ctx([]), ScanRaw(foreign));

      expect(d.outcome, Outcome.wrongEvent);
      expect(d.detail, contains('Yusuf & Maryam'),
          reason: 'this is the whole point of multi-key loading');
    });

    test('a genuine pass for the other leg is distinguished from a forgery',
        () {
      final abujaOnly = household();
      // Signed for this event, but absent from the Lagos leg's list.
      final d = decide(ctx([]), ScanRaw(tokenFor(abujaOnly)));

      expect(d.outcome, Outcome.wrongLeg);
      expect(d.outcome, isNot(Outcome.invalid));
    });

    test('an unreadable code offers name search rather than dead-ending', () {
      final d = decide(ctx([]), const ScanRaw('not-a-token'));

      expect(d.outcome, Outcome.invalid);
      expect(d.actions, contains('Search by name'),
          reason: 'usually a real guest with a broken screen');
    });

    test('a reissued pass reads as revoked, not invalid', () {
      final inv = household();
      final old = tokenFor(inv, version: 1);
      final d = decide(
        ctx([inv], keys: [
          EventKey(
            eventId: weddingKey.eventId,
            eventName: weddingKey.eventName,
            tokenVersion: 2,
            key: weddingKey.key,
          ),
        ]),
        ScanRaw(old),
      );

      expect(d.outcome, Outcome.revoked);
      expect(d.detail, contains('new link'));
    });
  });

  // -------------------------------------------------------------------------

  group('the RSVP gate', () {
    final cases = <(Rsvp, bool, Outcome)>[
      (Rsvp.pending, false, Outcome.admitted),
      (Rsvp.pending, true, Outcome.rsvpBlocked),
      (Rsvp.attending, true, Outcome.needsCount),
      (Rsvp.partial, true, Outcome.needsCount),
    ];

    for (final (rsvp, requireRsvp, expected) in cases) {
      test('rsvp=${rsvp.name}, required=$requireRsvp → ${expected.wire}', () {
        final inv =
            household(rsvp: rsvp, allowance: rsvp == Rsvp.pending ? 1 : 4);
        final d = decide(
          ctx([inv],
              policy: Policy(allowOverflow: true, requireRsvp: requireRsvp)),
          ScanRaw(tokenFor(inv)),
        );
        expect(d.outcome, expected);
      });
    }

    test('declining is refused whatever the policy says', () {
      final inv = household(rsvp: Rsvp.declined);
      final d = decide(
        ctx([inv],
            policy: const Policy(allowOverflow: true, requireRsvp: false)),
        ScanRaw(tokenFor(inv)),
      );
      expect(d.outcome, Outcome.rsvpDeclined);
    });

    test('override is offered only to ushers who hold the permission', () {
      final inv = household(rsvp: Rsvp.declined);

      final without = decide(ctx([inv]), ScanRaw(tokenFor(inv)));
      expect(without.actions, isNot(contains('Admit anyway')));

      final withPermission = decide(
        ctx([inv], canOverrideRsvp: true),
        ScanRaw(tokenFor(inv)),
      );
      expect(withPermission.actions, contains('Admit anyway'));
    });
  });

  // -------------------------------------------------------------------------

  group('manual check-in', () {
    test('skips every token check and is logged as manual', () {
      final inv = household(allowance: 1);
      final d = decide(ctx([inv]), ManualInput(inv.passId, requestedCount: 1));

      expect(d.outcome, Outcome.manual);
      expect(d.tone, Tone.admit);
      expect(d.admittedCount, 1);
    });

    test('a name that is not on the list offers a walk-in', () {
      final d = decide(ctx([]), ManualInput(randomUuid()));

      expect(d.outcome, Outcome.notFound);
      expect(d.actions, contains('Add walk-in'));
    });
  });

  // -------------------------------------------------------------------------

  group('invariants that must hold everywhere', () {
    List<Decision> everyScenario() {
      final good = household();
      final single = household(allowance: 1);
      final done = household(admitted: 4);
      final gone = household(revoked: true);
      final no = household(rsvp: Rsvp.declined);

      return [
        decide(ctx([single]), ScanRaw(tokenFor(single))),
        decide(ctx([good]), ScanRaw(tokenFor(good))),
        decide(ctx([good]), ScanRaw(tokenFor(good), requestedCount: 2)),
        decide(ctx([good]), ScanRaw(tokenFor(good), requestedCount: 9)),
        decide(
          ctx([done],
              policy: const Policy(allowOverflow: false, requireRsvp: false)),
          ScanRaw(tokenFor(done)),
        ),
        decide(ctx([gone]), ScanRaw(tokenFor(gone))),
        decide(ctx([no]), ScanRaw(tokenFor(no))),
        decide(ctx([]), const ScanRaw('rubbish')),
        decide(ctx([]), ManualInput(randomUuid())),
      ];
    }

    test('only green outcomes reset the camera by themselves', () {
      for (final d in everyScenario()) {
        if (d.autoReturnMs != null) {
          expect(
            d.tone == Tone.admit || d.outcome == Outcome.overflowAdmitted,
            isTrue,
            reason: '${d.outcome.wire} auto-returned but is not an admission',
          );
        }
      }
    });

    test('nothing is admitted on a refusal', () {
      const refusals = {
        Outcome.invalid,
        Outcome.wrongEvent,
        Outcome.wrongLeg,
        Outcome.revoked,
        Outcome.rsvpBlocked,
        Outcome.rsvpDeclined,
        Outcome.allowanceExhausted,
        Outcome.overflowBlocked,
        Outcome.notFound,
      };
      for (final d in everyScenario()) {
        if (refusals.contains(d.outcome)) {
          expect(d.admittedCount, 0);
        }
      }
    });

    test('every refusal is recorded — the organiser wants that report', () {
      for (final d in everyScenario()) {
        if (d.tone == Tone.deny || d.tone == Tone.hold) {
          expect(d.log, isTrue, reason: '${d.outcome.wire} was not logged');
        }
      }
    });

    test('only the count prompt goes unlogged', () {
      for (final d in everyScenario()) {
        if (!d.log) expect(d.outcome, Outcome.needsCount);
      }
    });

    test("no refusal leaks another household's identity", () {
      final foreign = issueToken(
        TokenPayload(
            passId: randomUuid(), eventId: otherWedding, tokenVersion: 1),
        otherKey.key,
      );
      final d = decide(ctx([]), ScanRaw(foreign));
      expect(d.invitation, isNull);
    });

    test('every outcome gives the usher something to read', () {
      for (final d in everyScenario()) {
        expect(d.headline, isNotEmpty);
      }
    });
  });

  // -------------------------------------------------------------------------

  group('two phones, both offline, same pass', () {
    test("each admits independently — reconciliation is the server's job", () {
      final shared = household(allowance: 4);

      // Neither device has seen the other's scan.
      final mainGate = decide(
        ctx([shared]),
        ScanRaw(tokenFor(shared), requestedCount: 3),
      );
      final sideGate = decide(
        ctx([shared]),
        ScanRaw(tokenFor(shared), requestedCount: 2),
      );

      expect(mainGate.admittedCount, 3);
      expect(sideGate.admittedCount, 2);

      // Five admitted against an allowance of four. Both rows are kept, the
      // server flags the overlap, and nobody already inside is thrown out.
      expect(mainGate.admittedCount + sideGate.admittedCount, 5);
      expect(mainGate.log && sideGate.log, isTrue);
    });
  });
}
